docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,235 @@
+# User-Defined Type Functions
+
+## Summary
+
+The new type solver for Luau supports a limited number of built-in type functions that enable developers to be expressive with their code. To further enhance this expressivity and the flexibility of the type system, this RFC proposes the design and implementation of user-defined type functions, a feature that will allow developers to define custom type functions.
+
+## Motivation
+
+The primary motivation for introducing user-defined type functions is to extend the flexibility of the type system and empower developers to create more expressive and type-safe libraries. Current limitations in the type system restrict custom type manipulations, which are crucial for creating advanced patterns and abstractions. By supporting user-defined type functions, we aim to:
+- Enable more precise type definitions that can capture complex relationships and constraints
+- Facilitate the creation of reusable and composable libraries
+- Enhance code safety and maintainability with better type-level programming
+
+The expected outcome is a more versatile type system that can adapt to a wider range of programming patterns, ultimately improving the Luau developer experience.
+
+## Design
+
+Type functions wil be defined with the following syntax:
+```luau
+type function f(...)
+    -- implementation of the type function
+end
+```
+
+For instance, the `rawget` type function can be written as:
+```luau
+type function rawget(tbl, prop)
+    if (not tbl:istable()) then
+        error("First parameter of rawget is not a table!") -- fails to reduce
+    end
+
+    return tbl:getprops()[prop:getvalue()]
+end
+
+type Person = {
+    name: string,
+    age: number
+}
+
+type ty = rawget<Person, "name"> -- ty = string
+```
+
+Type functions operate on two stages: type analysis and runtime. When calling type functions at the type level (e.g. annotating a variable as a type function), angle brackets must be used, but when calling them at the runtime level (e.g. calling other type functions within type functions), parenthesis must be used. Declarations of type functions use parenthesis because it defines the runtime operations on the runtime representation of types.
+
+For the first iteration, type functions will support a subset of the Luau language, mainly the constructs that do not allow type functions to run infinitely. Infinitely running type functions are problematic because they can halt the analysis from making further progress. For example, reducing this type function will recurse until a stack overflow occurs and cause the whole analysis to crash:
```

Comment:
"Cause the whole analysis to crash" is absolutely not true. The stack overflow is in a VM instance, and you can literally read it out as the error from the VM and report to the user that the type function stack overflowed.